bigquery_etl/cli/stage.py

```diff
@@ -0,0 +1,367 @@
+)
+@sql_dir_option
+@click.option(
+    "--dataset-suffix",
+    "--dataset_suffix",
+    help="Suffix appended to the deployed dataset",
+)
+@click.option(
+    "--update-references",
+    "--update_references",
+    default=True,
+    help="Update references to deployed artifacts with project and dataset artifacts have been deployed to.",
+)
+@click.option(
+    "--copy-sql-to-tmp-dir",
+    "--copy_sql_to_tmp_dir",
+    help="Copy existing SQL from the sql_dir to a temporary directory and apply updates there.",
+    default=False,
+    is_flag=True,
+)
+@click.pass_context
+def deploy(
+    ctx,
+    paths,
+    project_id,
+    sql_dir,
+    dataset_suffix,
+    update_references,
+    copy_sql_to_tmp_dir,
+):
+    """Deploy provided artifacts to destination project."""
+    if copy_sql_to_tmp_dir:
+        # copy SQL to a temporary directory
+        tmp_dir = Path(tempfile.mkdtemp())
+        tmp_dir.mkdir(parents=True, exist_ok=True)
+        shutil.copytree(sql_dir, tmp_dir)
+        sql_dir = tmp_dir / sql_dir.name
+
+    artifact_files = set()
+
+    # get SQL files for artifacts that are to be deployed
+    for path in paths:
+        artifact_files.update(
+            paths_matching_name_pattern(path, sql_dir, None, files=["*.sql", "*.py"])
+        )
+
+    # any dependencies need to be determined an deployed as well since the stage
+    # environment doesn't have access to the prod environment
+    artifact_files.update(_udf_dependencies(artifact_files))
+    artifact_files.update(_view_dependencies(artifact_files, sql_dir))
+
+    # update references of all deployed artifacts
+    # references needs to be set to the stage project and the new dataset identifier
+    if update_references:
+        _update_references(artifact_files, project_id, dataset_suffix, sql_dir)
+
+    updated_artifact_files = set()
+    # copy updated files locally to a folder representing the stage env project
+    for artifact_file in artifact_files:
+        project = artifact_file.parent.parent.parent.name
+        dataset = artifact_file.parent.parent.name
+        name = artifact_file.parent.name
+        test_path = TEST_DIR / project / dataset / name
+
+        if dataset_suffix:
+            dataset = f"{dataset}_{dataset_suffix}"
+
+        new_artifact_path = Path(sql_dir) / project_id / dataset / name
+        new_artifact_path.mkdir(parents=True, exist_ok=True)
+        shutil.copytree(artifact_file.parent, new_artifact_path, dirs_exist_ok=True)
+        updated_artifact_files.add(new_artifact_path / artifact_file.name)
+
+        # copy tests to the right structure
+        if test_path.exists():
+            shutil.copytree(
+                test_path, TEST_DIR / project_id / dataset / name, dirs_exist_ok=True
+            )
+            shutil.rmtree(test_path)
+
+    # remove artifacts from the "prod" folders
+    for artifact_file in artifact_files:
+        if artifact_file.parent.exists():
+            shutil.rmtree(artifact_file.parent)
```

Comment:
Does this remove stuff from `sql/` directory if `--copy_sql_to_tmp_dir` is not set?
If so then I'd be surprised by that if I ran this locally.

Reply:
Yes, it would. I guess we could have a flag to control this. However, if these files do not get deleted then all tests run on them (dryrun, view validation, etc) are likely going to fail since the referenced artifacts don't actually exist or the folder structure doesn't conform to the dataset and project names of the artifacts that got deployed to stage.